node/src/ledger/ledger_manager.rs

```diff
@@ -0,0 +1,515 @@
+                let res =
+                    ledger_ctx.staged_ledger_aux_and_pending_coinbase(ledger_hash, protocol_states);
+                Ok(LedgerResponse::LedgerAuxAndCoinbaseResult(res))
+            }
+            LedgerRequest::StagedLedgerDiffCreate {
+                pred_block,
+                won_slot,
+                coinbase_receiver,
+                completed_snarks,
+                supercharge_coinbase,
+            } => ledger_ctx
+                .staged_ledger_diff_create(
+                    &pred_block.lock().unwrap(),
+                    &won_slot.lock().unwrap(),
+                    &coinbase_receiver.lock().unwrap(),
+                    completed_snarks,
+                    supercharge_coinbase,
+                )
+                .map(LedgerResponse::StagedLedgerDiff),
+            LedgerRequest::StagedLedgerReconstruct {
+                snarked_ledger_hash,
+                parts,
+            } => {
+                ledger_ctx.staged_ledger_reconstruct(snarked_ledger_hash, parts);
+                Ok(LedgerResponse::Success)
+            }
+            LedgerRequest::StakeProofSparseLedger {
+                staking_ledger,
+                producer,
+                delegator,
+            } => {
+                let res = ledger_ctx.stake_proof_sparse_ledger(staking_ledger, producer, delegator);
+                Ok(LedgerResponse::SparseLedgerBase(res))
+            }
+        }
+    }
+}
+
+struct LedgerRequestWithChan {
+    request: LedgerRequest,
+    responder: oneshot::Sender<Result<LedgerResponse, String>>,
+}
+pub struct LedgerManager {
+    receiver: mpsc::UnboundedReceiver<LedgerRequestWithChan>,
+    sender: mpsc::UnboundedSender<LedgerRequestWithChan>,
+    ledger_ctx: LedgerCtx,
+}
+
+impl LedgerManager {
+    pub fn new(ledger_ctx: LedgerCtx) -> LedgerManager {
+        let (sender, receiver) = mpsc::unbounded_channel();
+        LedgerManager {
+            receiver,
+            sender,
+            ledger_ctx,
+        }
+    }
+
+    pub fn run(&mut self) {
+        tokio::runtime::Builder::new_current_thread()
+            .enable_all()
+            .build()
+            .unwrap()
+            .block_on(async {
+                loop {
+                    match self.receiver.recv().await {
+                        Some(LedgerRequestWithChan { request, responder }) => responder
+                            .send(request.handle(&mut self.ledger_ctx))
+                            .unwrap_or(()),
+                        None => {
+                            break;
+                        }
+                    }
+                }
+            });
+    }
+
+    fn call(&self, request: LedgerRequest) -> Result<LedgerResponse, String> {
+        let (responder, receiver) = oneshot::channel();
+        self.sender
+            .send(LedgerRequestWithChan { request, responder })
+            .unwrap();
+        result_join(receiver.blocking_recv())
+    }
+
+    pub fn accounts_set(
```

Comment:
Is there any advantage on wrapping the ledger address `Arc<Mutex<...>>` instead of just cloning it and passing it by-value?

Reply:
Not sure. I changed all the references in the old API to `Arc<Mutex<...>>`, because it was a simplest solution. Copying is sometimes costly and sometimes not possible at all. Admittedly I didn't checks these particular examples. I assumed that they were passed by reference previously for a reason and given that these `Arc`s aren't meant to live long (at most until the request is handled), I thought it the safest choice. I didn't consider the fact that wrapping in `Arc<Mutex<...>>` might require substantial changes to where these functions are called. I will review these refs and clone wherever possible.